File: demo/android/app/src/main/java/com/morph/markdown/demo/StreamingSimulator.kt
package com.morph.markdown.demo

import kotlin.math.min
import kotlin.random.Random

data class StreamChunk(
	val text: String,
	val delayMs: Long
)

object StreamingSimulator {
	private val random = Random(20260714)

	fun create(markdown: String): List<StreamChunk> {
		val out = mutableListOf<StreamChunk>()
		var index = 0
		while (index < markdown.length) {
			val next = nextEnd(markdown, index)
			val text = markdown.substring(index, next)
			out.add(StreamChunk(text, delayFor(text)))
			index = next
		}
		return out
	}

	private fun nextEnd(markdown: String, start: Int): Int {
		val target = start + nextTokenSize(markdown, start)
		val bounded = min(target, markdown.length)
		return extendToNaturalBoundary(markdown, start, bounded)
	}

	private fun nextTokenSize(markdown: String, start: Int): Int {
		val ch = markdown[start]
		if (ch == '\n') return if (markdown.getOrNull(start + 1) == '\n') 2 else 1
		return when (random.nextInt(100)) {
			in 0..52 -> random.nextInt(1, 5)
			in 53..82 -> random.nextInt(5, 11)
			in 83..94 -> random.nextInt(11, 20)
			else -> random.nextInt(20, 36)
		}
	}

	private fun extendToNaturalBoundary(markdown: String, start: Int, end: Int): Int {
		if (end >= markdown.length) return markdown.length
		if (end == start) return min(start + 1, markdown.length)
		if (markdown[end - 1].isWhitespace()) return end
		val next = markdown[end]
		if (next.isWhitespace() || isPunctuation(next)) return end + 1
		return end
	}

	private fun delayFor(text: String): Long {
		val base = when {
			text.contains("\n\n") -> random.nextLong(460, 920)
			text.any { it == '\n' } -> random.nextLong(220, 520)
			text.any { isStrongPunctuation(it) } -> random.nextLong(180, 420)
			text.length <= 3 -> random.nextLong(24, 90)
			text.length <= 10 -> random.nextLong(45, 140)
			else -> random.nextLong(80, 210)
		}
		return withOccasionalPause(base)
	}

	private fun withOccasionalPause(base: Long): Long {
		return when (random.nextInt(100)) {
			in 0..4 -> base + random.nextLong(420, 950)
			in 5..16 -> base + random.nextLong(120, 360)
			else -> base
		}
	}

	private fun isPunctuation(ch: Char): Boolean {
		return ch in punctuation
	}

	private fun isStrongPunctuation(ch: Char): Boolean {
		return ch in strongPunctuation
	}

	private val punctuation = setOf(
		'.', ',', ':', ';', '!', '?', ')', ']', '}', '>', '|',
		'。', '，', '：', '；', '！', '？', '）', '】', '》', '、'
	)

	private val strongPunctuation = setOf(
		'.', '!', '?', ';', '。', '！', '？', '；'
	)
}

File: demo/ios/MorphMarkdownDemo/MorphMarkdownDemo/StreamingSimulator.swift
import Foundation

struct StreamChunk {
	let text: String
	let delay: TimeInterval
}

enum StreamingSimulator {
	static func create(markdown: String) -> [StreamChunk] {
		var generator = SeededGenerator(seed: 20260714)
		var out: [StreamChunk] = []
		var index = markdown.startIndex
		while index < markdown.endIndex {
			let next = nextIndex(in: markdown, from: index, generator: &generator)
			let text = String(markdown[index..<next])
			out.append(StreamChunk(text: text, delay: delay(for: text, generator: &generator)))
			index = next
		}
		return out
	}

	private static func nextIndex(in text: String, from start: String.Index, generator: inout SeededGenerator) -> String.Index {
		let count = nextTokenSize(in: text, from: start, generator: &generator)
		let raw = text.index(start, offsetBy: count, limitedBy: text.endIndex) ?? text.endIndex
		return extendToNaturalBoundary(in: text, start: start, end: raw)
	}

	private static func nextTokenSize(in text: String, from start: String.Index, generator: inout SeededGenerator) -> Int {
		if text[start] == "\n" {
			let next = text.index(after: start)
			return next < text.endIndex && text[next] == "\n" ? 2 : 1
		}
		switch generator.nextInt(upperBound: 100) {
		case 0...52: return generator.nextInt(in: 1...4)
		case 53...82: return generator.nextInt(in: 5...10)
		case 83...94: return generator.nextInt(in: 11...19)
		default: return generator.nextInt(in: 20...35)
		}
	}

	private static func extendToNaturalBoundary(in text: String, start: String.Index, end: String.Index) -> String.Index {
		if end >= text.endIndex {
			return text.endIndex
		}
		if end == start {
			return text.index(after: start)
		}
		let previous = text.index(before: end)
		if text[previous].isWhitespace {
			return end
		}
		return text[end].isWhitespace || punctuation.contains(text[end]) ? text.index(after: end) : end
	}

	private static func delay(for text: String, generator: inout SeededGenerator) -> TimeInterval {
		let base: Int
		if text.contains("\n\n") {
			base = generator.nextInt(in: 460...920)
		} else if text.contains("\n") {
			base = generator.nextInt(in: 220...520)
		} else if text.contains(where: { strongPunctuation.contains($0) }) {
			base = generator.nextInt(in: 180...420)
		} else if text.count <= 3 {
			base = generator.nextInt(in: 24...90)
		} else if text.count <= 10 {
			base = generator.nextInt(in: 45...140)
		} else {
			base = generator.nextInt(in: 80...210)
		}
		return TimeInterval(base + occasionalPause(generator: &generator)) / 1000
	}

	private static func occasionalPause(generator: inout SeededGenerator) -> Int {
		switch generator.nextInt(upperBound: 100) {
		case 0...4: return generator.nextInt(in: 420...950)
		case 5...16: return generator.nextInt(in: 120...360)
		default: return 0
		}
	}

	private static let punctuation = Set(".，,:;!?)]}>|。：；！？）】》、")
	private static let strongPunctuation = Set(".!?;。！？；")
}

private struct SeededGenerator {
	private var state: UInt64

	init(seed: UInt64) {
		state = seed
	}

	mutating func nextInt(upperBound: Int) -> Int {
		return Int(next() % UInt64(upperBound))
	}

	mutating func nextInt(in range: ClosedRange<Int>) -> Int {
		return range.lowerBound + nextInt(upperBound: range.upperBound - range.lowerBound + 1)
	}

	private mutating func next() -> UInt64 {
		state = state &* 6364136223846793005 &+ 1442695040888963407
		return state >> 32
	}
}
